### DISCUSSION TITLE
What development environment do you use?

### DISCUSSION THREAD

**Discussion by Matyrobbrt** [neoforged/NeoForge#165] [Polls]
A question for NeoForge mod developers.

**Comment by Hipposgrumm** [comment]
Maybe we should say that if you vote Other you should say what you use in the comments.

Also, you can change your answer by clicking the Hide Results button.

**Reply by ghost** [reply]
I use notepad

**Reply by gigaherz** [reply]
I commend your choice to use the most arduous path.

**Comment by williewillus** [comment]
Where is the option for "begrudgingly uses Intellij, but would use anything else if possible (mainly the eclipse lsp + emacs)"

**Reply by dphaldes** [reply]
The day I figure how to get neovim to work well with java is the day I say goodbye to intellij :P

**Comment by williewillus** [comment]
If this is to gauge whether to support non intellij environments (so eclipse, basically), I'd strongly recommend keeping at least baseline working support for it. Dropping it and locking everyone into intellij wouldn't be good.

**Reply by esotericist** [reply]
i, at least, can't function at all in intellij or eclipse. vs code is pretty much the only ide i can cope with.

**Reply by PickleOnAString** [reply]
vscode 4 life

**Comment by wgaylord** [comment]
I use IntelliJ mostly due to not trusting Eclipse anymore since like 1.5 Minecraft times when I had it literally forget that a src folder existed on every project I opened. Never had any issues with IntelliJ.

**Comment by ShroomAgent27** [comment]
As long as eclipse can properly build the classpath I'm fine. I always use the command line for gradle tasks

**Comment by ChampionAsh5357** [comment]
I typically use all three, but the most would be IntelliJ and Eclipse. I do use VSCode when using languages other than Java. As for the others, Gradle integration is always an interesting time, but it works occasionally.

**Comment by Sinhika** [comment]
I tried IntelliJ once, but that was during the time it was buggy and not importing source code libs properly--and you can't do any serious modding if you can't see the MC and Forge sources. Turned me off the product completely.

**Comment by TheAGamePlayer** [comment]
Eclipse is the obvious choice for me atm. Works mostly flawlessly however is slightly buggy on Ubuntu (Less buggy on other distros). Out of principle I've avoided InteliiJ due to it's ridiculous pay model being a leech to my wallet & a thorn in my side outside of coding my Minecraft mods. I've strongly been debating using Visual Studio Code, due to Visual Studio being very nice to work with my Rust, C, & C++ coding (with the exception of requiring a Microsoft account login). I'd only make the change if I found Eclipse to still have annoying bugs when I switch to either Linux Mint or Fedora.

**Reply by Hipposgrumm** [reply]
You know that IntelliJ has a *free* version, right? 😅

**Reply by TelepathicGrunt** [reply]
In addition to the free community version available, you are able to get the ultimate version for free if you have open source projects: https://www.jetbrains.com/community/opensource/#support

**Reply by TheAGamePlayer** [reply]
There's a reason I said "outside of coding my Minecraft mods" for the paid version :)
It has however been a year, so I will consider giving the community version a go before I settle on using Eclipse again.

**Comment by bluebear94** [comment]
I currently use the community edition of IntelliJ IDEA, but on my old laptop, I switched to Eclipse because IDEA was using too many resources.

**Comment by CreativeMD** [comment]
Personally there is one feature in particular I don't want to miss. That's why I use eclipse. The ability to start the application with lots and lots of compiler errors present. That allows testing stuff before everything else is ready. Besides that eclipse is a bit clunky, but you learn to live with them. Besides that it has a few more good features and is not missing anything essential.

**Reply by PlatinPython** [reply]
🤨 How tf does that work? Like how would you run a program that doesn't even compile.

**Reply by gigaherz** [reply]
Eclipse inserts an exception throw in parts of the code it couldn't compile, and then moves on with the compilation. It's horrible but some people like it.

**Reply by Technici4n** [reply]
You can use ECJ with IntelliJ if you want:

![image](https://github.com/neoforged/NeoForge/assets/13494793/39f0d3b5-0e94-4679-ae8a-597bb299da5f)

**Comment by lukebemish** [comment]
I started out with Eclipse and use VSCode extensively for non-java work, but for modding I have switched entirely to IntelliJ. Of the three it's the only one I've found that has actual gradle support, especially when non-trivial build systems start to pop up. I've had too many projects in the past that eclipse just wouldn't work with in the slightest.

That said, keeping NeoGradle support for eclipse is probably good, as IntelliJ (VSCode too to a lesser degree) is an enormous resource hog, and if I was working on more of a potato than I do now I'd almost certainly be trying out eclipse again. However, I certainly wouldn't hold up features based on it (no clue if this is actually an issue) - if eclipse can't properly support gradle, that's their issue

**Comment by RaveTr** [comment]
I typically use both, but since I use Eclipse more it's the vote I put in.

**Comment by PrincessOfEvil** [comment]
I usually use Eclipse, for some reason IDEA craps out on me if i use it for too long.
(I also use notepad, sometimes, but i don't think NF has a plugin for it.)

**Comment by IriDark** [comment]
Notepad++, tbh it's crazy.

**Reply by AllenSeitz** [reply]
Dropping support for Eclipse is the only thing that would send me back to LexForge. I hate that you're even asking the question. (Because lets face it, that's the thinly-veiled question you're really asking here.)

**Reply by AllenSeitz** [reply]
Meant that as a reply to the poll, not to you, sorry.

Keep rockin' Notepad++ heck yeah!

**Reply by TelepathicGrunt** [reply]
@AllenSeitz It's not about dropping support. And it is not a thinly-veiled question. It's more about how Neoforge is composed of volunteers with only so much time and effort they can dedicate to the toolchains. The question is to gauge just how much effort should be poured into high quality support for other IDEs or if it should be just basic support, or "best effort but not RC blocking". Not so much dropping IDEs entirely. 

Also, those who work on the less popular IDEs could come in and join Neoforge and help further improve the support for those less used IDEs as well so the Neo team can focus on other work.

**Reply by AllenSeitz** [reply]
Thank you for your reply! I'm glad to hear it and I'm sorry I was rude. I panicked because Eclipse is doing about as well as VSCode in the poll.

But there's probably no chance of me being able to contribute the Eclipse makefile thingums that I need? If I knew how to do the thing then I wouldn't be dependent on other people to provide the thing for me. I'm just a customer here.

**Reply by lukebemish** [reply]
The useful information that comes out of this poll is going to be something like "the effort put in to support eclipse should probably be similar to what's put in to support VSCode, and IntelliJ support should be a huge priority" or that sort of thing; given that several neoforge developers use eclipse I think we can safely assume it's not getting dropped and will keep working well enough

**Comment by LoneStarrr** [comment]
I Begrudgingly use IntelliJ, even if it's overly complex for a very casual developer. Try figuring out how to manage multiple projects in a single workspace. At least that was always easy with Eclipse, in IntelliJ it seems to be a half-hearted afterthought.

**Reply by gigaherz** [reply]
That's simply because Intellij is *not* built around the idea of a multi-project workspace, despite people's efforts to make it a thing.
In Eclipse, people sometimes will put all their modding projects in one workspace and then work on whichever they feel like, and sometimes use one mod as a dependency of another. You are not expected to work with intellij like that. 
Intellij is built around an IDE instance having *one* project, optionally with sub-projects (modules) nested inside it. So, if you have multiple modding projects, you just open each one in a separate window, or close the previous one first if you don't want multiple windows. And yes, that means managing a library which is being used by multiple mod projects at the same time, is a bit more annoying.

**Comment by BoogieMonster1O1** [comment]
I usually use emacs but the mcdev plugin with intellij makes mixins so much easier to work with.

**Comment by HeatherComputer** [comment]
Proud vscode user. works inifnitely better for me than intellij has ever hoped to, and it works great with other languages as a bonus.

**Reply by ZedDevStuff** [reply]
How do you even do that? Tried VScode for (Minecraft modding) a while ago and couldn't get anything to work by following the instructions

**Reply by HeatherComputer** [reply]
If you're newer than 1.12, there's a simple `genVSCodeRuns` task included with neo/forge and fabric mdks (not always that exact name of course). Use that instead of the gen idea runs task, and install the java pack that vscode prompts you to get - that's about it really. 

Things can have quirks - for debugging specifically in neoforge for example, you'd want to add a gradle task - but that's about it.


<details>
<summary>Gradle Task</summary>

```groovy
task copyVSCodeResources (dependsOn: 'processResources', type: Copy) {
    from 'build/resources'
    into 'bin/'
}
```
</details>

And then add `"preLaunchTask": "gradle: copyVSCodeResources"` to your generated runs. This shouldn't be needed for other fabric or forge MDKs.


Of course, making the runs manually is doable - but I've no reason to do that when the MDKs bundle a run gen task. If that stops being bundled, I'll have to make a task of my own haha

**Comment by TheOriginalGolem** [comment]
I use Intellij with the minecraft development plugin(has alot of really useful tools)

**Reply by Acenio** [reply]
It does and would be nice if the dev of that plugin added/switched to neo

**Comment by sekelsta** [comment]
Other: [Pluma](https://en.wikipedia.org/wiki/Pluma_(text_editor))
